### PR TITLE
improved testing

### PR DESCRIPTION
**Pull request by wakamex**
**major changes**:
- fix get_max_pt_short
- print market_step only if executed trades
- pass kwargs to agents to allow arbitrary parameters passed in testing
- add testing to verify that custom parameters are working as intended
- make deleting log file optional with a parameter

**fixes**:
- fix bug that initialized mint_time to 0 instead of None which resulted in wrong time_remaining calculation
- fix close_short accounting not receiving sale proceeds
- revert previous incorrect attempts to fix max short
- add handler close statements to avoid file permission errors in windows

**minor changes**:
- remove float_to_string since it's not useful for logging (only for printing to terminal)
- remove fees_paid since it isn't needed
- remove effective_price since it's informative only, can be calculated post-sim
- rename time_remaining to stretched_time_remaining to clarify what should be passed in
- additional comments to clarify what units are being passed in
- rename amount_to_lp to base_to_lp
- few additional logging statements and clarifications

### CODE REVIEW

File: src/elfpy/agent.py [hunk 1]
```diff
@@ -3,10 +3,10 @@
 """
 
 import logging
-
```
Comment (dpaiton) [inline src/elfpy/agent.py:6]: newline between stdlib & external imports is a good idea (black should enforce this?)

File: src/elfpy/agent.py [hunk 2]
```diff
@@ -27,9 +32,14 @@ def __init__(self, market: Market, rng: Generator, wallet_address: int, budget:
         # TODO: remove this, wallet_address is a property of wallet, not the agent
         self.wallet_address: int = wallet_address
         self.budget: float = budget
+        self.verbose: bool = False if verbose is None else verbose
```
Comment (dpaiton) [inline src/elfpy/agent.py:35]: We've gotten rid of verbose in the logging PR.

**Comment by dpaiton** [comment]
> pass kwargs to agents to allow arbitrary parameters passed in testing

Why is this a better pattern than assigning parameters in the config?

For example, we set simulation configs in the simulator by modifying the config object directly when it is called. For example,  in [test_sim](https://github.com/element-fi/elf-simulations/blob/main/tests/test_sim.py#L44) :
```
        simulator = YieldSimulator(
            Config(
                market=MarketConfig(),
                amm=AMMConfig(pricing_model_name="Hyperdrive"),
                simulator=SimulatorConfig(
                    num_trading_days=10,
                    num_blocks_per_day=10,
                    logging_level=logging.INFO,
                ),
            )
        )
```

For the user, we can do the same. Set user_list = [] on init, then add our own with args passed. e.g.:
```
simulator = YieldSimulator(config)
simulator.reset_rng(
    np.random.default_rng(simulator.config.simulator.random_seed)
)
simulator.set_random_variables()
simulator.setup_simulated_entities({"user_policies": []}) # no users at first
simulator.agents.update(
    {1: CustomUser(
            market=simulator.market,
            rng=simulator.rng,
            wallet_address=1
        )
    }
)
simulator.run_simulation()
```

CustomUser would be specified elsewhere, e.g. from a dictionary in the notebook or a toml file. 

This functionality already exists in main. So I'm thinking it's not really worth adding an alternative user pattern in your PR. I'm also weary because combining `setattr` with `kwargs` as you do is hard to parse. I didn't find a great explanation from a quick Google, but these two links get at the idea: [stack overflow answer](https://stackoverflow.com/a/13982449) and [some other weird forum answer](https://teamtreehouse.com/community/is-using-setattr-with-kwargs-a-good-idea). It's not like a terrible thing to do, but I'd rather avoid it if we can.

File: src/elfpy/agent.py [hunk 3]
```diff
@@ -18,7 +18,12 @@ class Agent:
     date value is an inte with how many tokens they have for that date
     """
 
-    def __init__(self, market: Market, rng: Generator, wallet_address: int, budget: float):
+    # pylint: disable=too-many-instance-attributes
```
Comment (dpaiton) [inline src/elfpy/agent.py:21]: If you're going to add pylint disables, please also add descriptions of the problem and a path to fix them. We should avoid disabling these if possible.

File: src/elfpy/agent.py [hunk 4]
```diff
@@ -64,8 +74,22 @@ def get_max_pt_short(self) -> float:
         """
         if self.market.share_reserves == 0:
             return 0
-        max_pt_short = self.market.share_reserves * self.market.share_price / self.market.get_spot_price()
-        return max_pt_short
+        time_remaining = time_utils.get_yearfrac_remaining(
```
Comment (dpaiton) [inline src/elfpy/agent.py:77]: Remove the TODO comment above as well please!

File: src/elfpy/agent.py [hunk 5]
```diff
@@ -64,8 +74,22 @@ def get_max_pt_short(self) -> float:
...
+            self.market.time, self.market.time, self.market.token_duration
```
Comment (dpaiton) [inline src/elfpy/agent.py:78]: can you name these args, so that they're more clear. Since the values are the same.

File: src/elfpy/agent.py [hunk 6]
```diff
@@ -109,7 +133,7 @@ def update_wallet(self, wallet_deltas: Wallet) -> None:
             if key in ["base_in_wallet", "lp_in_wallet", "fees_paid"]:
                 if value_or_dict != 0 or self.wallet[key] != 0:
                     logging.debug(
-                        "agent #%g %s pre-trade = %.0g\npost-trade = %1g\ndelta = %1g",
+                        "agent %03.0f %s pre-trade = %.0g\npost-trade = %1g\ndelta = %1g",
```
Comment (dpaiton) [inline src/elfpy/agent.py:136]: Why did you also remove the `#`?

File: src/elfpy/simulators.py [hunk 7]
```diff
@@ -347,18 +362,25 @@ def collect_and_execute_trades(self, last_block_in_sim=False):
...
                     [key for key in self.agents if key > 0]  # exclude init_lp before shuffling
                 )
                 wallet_ids = np.append(wallet_ids, 0)  # add init_lp so that they're always last
-            else:
+            else:  # include init_lp only on the last block, to let it unwind
                 wallet_ids = self.rng.permutation(list(self.agents))
+        else:  # we are in a deterministic mode
+            # reverse the list excluding 0 (init_lp)
+            wallet_ids = [key for key in self.agents if key > 0][::-1]
+            if last_block_in_sim:  # prepend init_lp to the list
```
Comment (dpaiton) [inline src/elfpy/simulators.py:382]: Why is the init_lp excluded for other blocks?

File: src/elfpy/simulators.py [hunk 8]
```diff
@@ -347,18 +362,25 @@ def collect_and_execute_trades(self, last_block_in_sim=False):
         """Get trades from the agent list, execute them, and update states"""
         if not isinstance(self.market, Market):
             raise ValueError("market not defined")
+        number_of_executed_trades = 0
         # TODO: This is a HACK to prevent the initial LPer from rugging other agents.
         # The initial LPer should be able to remove their liquidity and any open shorts can still be closed.
         # But right now, if the LPer removes liquidity while shorts are open,
         # then closing the shorts results in an error (share_reserves == 0).
+        wallet_ids = [key for key in self.agents if key > 0]  # exclude init_lp before shuffling
```
Comment (dpaiton) [inline src/elfpy/simulators.py:370]: Is this needed? I don't think it's used.

File: src/elfpy/simulators.py [hunk 9]
```diff
@@ -376,10 +398,14 @@ def collect_and_execute_trades(self, last_block_in_sim=False):
                 agent.log_status_report()
                 self.update_analysis_dict()
                 self.run_trade_number += 1
+                number_of_executed_trades += 1
+        if number_of_executed_trades > 0:
+            logging.debug("executed %f trades at %s", number_of_executed_trades, self.market.get_market_state_string())
```
Comment (dpaiton) [inline src/elfpy/simulators.py:403]: This adds a debug print for every trade, which will generate a massive amount of outputs. Maybe we should introduce a level higher than debug to warn users that they're getting hella outputs? It seems like there would be good argument for a debug output that is still viable with more than a few days, blocks per day, and trades per block.

File: src/elfpy/strategies/single_lp.py [hunk 10]
```diff
@@ -10,29 +10,14 @@
...
-        """call basic policy init then add custom stuff"""
-        self.amount_to_lp = 100
-        super().__init__(
-            market=market,
-            rng=rng,
-            wallet_address=wallet_address,
-            budget=budget,
-        )
+    "single LP: only has one LP open at a time"
+    amount_to_lp = 100
```
Comment (dpaiton) [inline src/elfpy/strategies/single_lp.py:14]: This is now a static element that belongs to the class. Before, the variable was an object member variable (belongs to `self`). The difference is whether you want the variable to exist even if the class is not instantiated. What you've done here is not necessarily _wrong_, but it is conveying the wrong message. We want `amount_to_lp` to be belonged by some instantiated user. e.g.
```
from elfpy.strategies.single_lp import Policy
alice = Policy()
alice.amount_to_lp = 100
bob = Policy()
bob.amount_to_lp = 1000
```

so in this example `amount_to_lp` belongs to alice or bob, not to Policy. 

Long story short, you should put this back to how it was.

File: src/elfpy/strategies/single_lp.py [hunk 11]
```diff
@@ -10,29 +10,14 @@
...
-        implement user strategy
-        LP if you can, but only do it once
-        """
-        action_list = []
         has_lp = self.wallet.lp_in_wallet > 0
         can_lp = self.wallet.base_in_wallet >= self.amount_to_lp
         if can_lp and not has_lp:
-            action_list.append(self.create_agent_action(action_type="add_liquidity", trade_amount=self.amount_to_lp))
-        return action_list
+            self.action_list.update(
```
Comment (dpaiton) [inline src/elfpy/strategies/single_lp.py:20]: It appears as though you deleted the docstring. Please put it back. Also, I think the action_list should not be a member variable of agent (mentioned in an earlier comment).

File: tests/test_logging.py [hunk 12]
```diff
@@ -60,6 +60,7 @@ def test_logging(self):
             simulator.setup_simulated_entities(override_dict)
             simulator.run_simulation()
             self.assertLogs(level=level)
+            handler.close()
```
Comment (dpaiton) [inline tests/test_logging.py:63]: Why did you need to add this? I think I was getting errors from not having it, but I'm curious what lead you to figure it out?

File: tests/test_trade.py [hunk 13]
```diff
@@ -122,3 +123,8 @@ def test_single_short(self):
     def test_base_lps(self):
         """Tests base LP setups"""
         self.run_base_lp_test(user_policies=["single_lp"], config_file="config/example_config.toml")
+
+
+if __name__ == "__main__":
```
Comment (dpaiton) [inline tests/test_trade.py:128]: I don't think we should commit tests with options to run as a script

File: tests/utils/test_time_utils.py [hunk 14]
```diff
@@ -23,10 +23,13 @@ def test_current_datetime(self):
 
         now = datetime.datetime.now(pytz.timezone("Etc/GMT-0"))
         test_time = time_utils.current_datetime()
+        time_delta = datetime.timedelta(milliseconds=100)
```
Comment (dpaiton) [inline tests/utils/test_time_utils.py:26]: This is testing something different now, right? Why did you make the change?

File: src/elfpy/agent.py (hunk 1 above)
Comment (wakamex) [reply src/elfpy/agent.py:6]: looks like it doesn't, but i can add it

**Comment by wakamex** [comment]
> Why is this a better pattern than assigning parameters in the config?
the only reason would be if it's easier for the user (which it may not be)

> CustomUser would be specified elsewhere, e.g. from a dictionary in the notebook or a toml file.
this sounds like additional boilerplate required for the user the understand and use. whereas adding `:parameter=value` inside a string they're already passing in, seems easier for the user?

> This functionality already exists in main.
it doesn't exist for agents, which is why i added this

> So I'm thinking it's not really worth adding an alternative user pattern in your PR
agree we shouldn't have two methods for the same thing

> I'm also weary because combining `setattr` with `kwargs` as you do is hard to parse.
it's definitely a bit hackish and hard to read. this is necessary to allow custom parameters to be created at init-time. related to this usage, i found the need to explicitly declare all member variables before this loop, to avoid making typos and setting a non-existent member variable that isn't going to be used (i made this mistake). that's messy, requiring each member variable to be declared here, 1 line for each variable. a cleaner version of that is using __slots__ to define all member variables ahead of time, in 1 line!!! while also enforcing stricter use of member variables (i assume you're a fan 😏). check this out:
```
    __slots__ = ["market", "rng", "wallet_address", "budget", "last_update_spend", "product_of_time_and_base",
                 "wallet", "action_list", "base_to_lp", "pt_to_short"]
```
making type on custom parameter definition in test_sequential.py: 
`ValueError: ERROR: Attempted to manually override non-existent Agent attribute based_to_lp`
making type trying to set a new member variable without running it by Dylan (he owns __slots__):
`haven't gotten this working yet, but my reading tells me it's possible` based on this: [https://wiki.python.org/moin/UsingSlots](https://wiki.python.org/moin/UsingSlots)
```
x1.slot1 = 1

    ---------------------------------------------------------------------------

    AttributeError                            Traceback (most recent call last)

    Input In [3], in <module>
    ----> 1 x1.slot1 = 1


    AttributeError: 'Example' object has no attribute 'slot1'
```

> Also I think the regular expression code added to the simulator would need to be pretty thoroughly tested to avoid weird inputs causing errors.
it's only used to chain together two splits. so it's not running an open-ended regular expression. just custom splitting. can probably be done without the regular expression library.

File: src/elfpy/pricing_models.py [hunk 15]
```diff
@@ -931,7 +931,7 @@ def calc_in_given_out(
         bond_reserves: float,
         token_in: TokenType,
         fee_percent: float,
-        time_remaining: float,
+        stretched_time_remaining: float,
```
Comment (dpaiton) [inline src/elfpy/pricing_models.py:934]: Good call -- I think this var name makes more sense.

File: src/elfpy/pricing_models.py [hunk 16]
```diff
@@ -1006,32 +1006,32 @@ def calc_in_given_out(
...
         # We precompute the YieldSpace constant k using the current reserves and
         # share price:
         #
         # k = (c / μ) * (μ * z)**(1 - t) + (2y + cz)**(1 - t)
         k = price_utils.calc_k_const(share_reserves, bond_reserves, share_price, init_share_price, time_elapsed)
         if token_in == "base":
-            in_reserves = share_reserves
-            out_reserves = bond_reserves + total_reserves
-            d_bonds = out
+            in_reserves = share_reserves  # in is in units of shares
```
Comment (dpaiton) [inline src/elfpy/pricing_models.py:1032]: Thanks for adding units!

File: src/elfpy/pricing_models.py [hunk 17]
```diff
@@ -1322,7 +1323,16 @@ def calc_out_given_in(
             #
             # fee = (1 - p) * φ * d_y
             fee = (1 - spot_price) * fee_percent * d_bonds
+
+            # To get the amount paid with fees, subtract the fee from the
```
Comment (dpaiton) [inline src/elfpy/pricing_models.py:1327]: Super clear -- thanks for adding this

Comment (dpaiton) [inline src/elfpy/pricing_models.py:1327]: consistent commenting 🤌

File: src/elfpy/simulators.py [hunk 18]
```diff
@@ -280,11 +280,26 @@ def setup_simulated_entities(self, override_dict=None):
             self.agents = {}
         self.market.log_market_step_string()
         # continue adding other users
-        for policy_number, policy_name in enumerate(self.config.simulator.user_policies):
+        for policy_number, policy_instruction in enumerate(self.config.simulator.user_policies):
+            if ":" in policy_instruction:  # we have custom parameters
```
Comment (dpaiton) [inline src/elfpy/simulators.py:284]: This looks a lot cleaner. I think it would be a good idea to also add a check function that receives the custom parameter and verifies the format with asserts. This way we provide more useful error messaging in situations where the custom parameters are formatted incorrectly or trying to do something that the user can't.

File: src/elfpy/simulators.py [hunk 19]
```diff
@@ -347,18 +362,25 @@ def collect_and_execute_trades(self, last_block_in_sim=False):
...
                     [key for key in self.agents if key > 0]  # exclude init_lp before shuffling
                 )
                 wallet_ids = np.append(wallet_ids, 0)  # add init_lp so that they're always last
-            else:
+            else:  # include init_lp only on the last block, to let it unwind
                 wallet_ids = self.rng.permutation(list(self.agents))
+        else:  # we are in a deterministic mode
+            # reverse the list excluding 0 (init_lp)
+            wallet_ids = [key for key in self.agents if key > 0][::-1]
+            if self.config.simulator.init_lp and last_block_in_sim:  # prepend init_lp to the list
```
Comment (dpaiton) [inline src/elfpy/simulators.py:382]: My understanding is that forcing init_lp to be last was a hack to ensure that the end-of-simulation accounting worked out and the printing didn't break. afaik, there's no hard reason why the init_lp agent has to stay in the simulation until the end, or has to be last in the closing trades.

If this is correct, do you think it's possible to update this PR to remedy the situation and remove the hack? If this is not correct, and the ordering is necessary, can you add a comment explaining why?

File: src/elfpy/strategies/single_lp.py [hunk 20]
```diff
@@ -6,33 +6,29 @@
...
-    def __init__(self, market, rng, wallet_address, budget=1000):
-        """call basic policy init then add custom stuff"""
-        self.amount_to_lp = 100
-        super().__init__(
-            market=market,
-            rng=rng,
-            wallet_address=wallet_address,
-            budget=budget,
-        )
+    "single LP: only has one LP open at a time"
```
Comment (dpaiton) [inline src/elfpy/strategies/single_lp.py:15]: This is now a static element that belongs to the class. Before, the variable was an object member variable (belongs to `self`). The difference is whether you want the variable to exist even if the class is not instantiated. We want `amount_to_lp` to be owned by some instantiated user. e.g.
```
from elfpy.strategies.single_lp import Policy
alice = Policy()
alice.amount_to_lp = 100
bob = Policy()
bob.amount_to_lp = 500
```

so in this example `amount_to_lp` belongs to alice or bob, not to Policy.  As you have it above, it belongs to the policy class.

tl;dr -- put it back in the init

File: src/elfpy/strategies/single_short.py [hunk 21]
```diff
@@ -4,29 +4,33 @@
...
-    simple short
-    only has one long open at a time
-    """
+    """single short: only has one long open at a time"""
 
-    def __init__(self, market, rng, wallet_address, budget=100):
-        """call basic policy init then add custom stuff"""
-        self.pt_to_short = 100
-        super().__init__(market, rng, wallet_address, budget)
+    pt_to_short = 100
```
Comment (dpaiton) [inline src/elfpy/strategies/single_short.py:15]: See above; put `__init__` back

File: src/elfpy/strategies/single_lp.py [hunk 22]
```diff
@@ -6,33 +6,29 @@
...
     def action(self):
-        """
-        implement user strategy
-        LP if you can, but only do it once
-        """
         action_list = []
         has_lp = self.wallet.lp_in_wallet > 0
-        can_lp = self.wallet.base_in_wallet >= self.amount_to_lp
+        can_lp = self.wallet.base_in_wallet >= self.base_to_lp
+        logging.info(
```
Comment (dpaiton) [inline src/elfpy/strategies/single_lp.py:22]: This should probably be logging.debug since it prints every trade

File: src/elfpy/strategies/single_short.py [hunk 23]
```diff
@@ -4,29 +4,33 @@
...
 
     def action(self):
-        """
```
Comment (dpaiton) [inline src/elfpy/strategies/single_short.py:22]: Every function, class, and python file needs a docstring

File: src/elfpy/wallet.py [hunk 24]
```diff
@@ -5,7 +5,7 @@
 from dataclasses import dataclass
 from dataclasses import field
 
-from elfpy.utils.outputs import float_to_string
+from elfpy.utils.price import calc_apr_from_spot_price
```
Comment (dpaiton) [inline src/elfpy/wallet.py:8]: Is this getting used? I can't find where

File: tests/test_trade.py [hunk 25]
```diff
@@ -39,7 +39,7 @@ def setup_logging():
             handler,
         ]
 
-    def run_base_trade_test(self, user_policies, config_file, additional_overrides=None):
+    def run_base_trade_test(self, user_policies, config_file, delete_log_file=True, additional_overrides=None):
```
Comment (dpaiton) [inline tests/test_trade.py:42]: Good idea, this is better than commenting.

File: tests/test_trade.py [hunk 26]
```diff
@@ -90,18 +93,21 @@ def run_base_lp_test(self, user_policies, config_file, additional_overrides=None
...
         market_apr = simulator.market.get_rate()
-        # check that apr is within a 0.1% of the target
-        assert np.allclose(
-            market_apr, target_pool_apr, atol=0.001
-        ), f"test_trade.run_base_lp_test: ERROR: {target_pool_apr=} does not equal {market_apr=}"
-        # check that the liquidity is within 5% of the target
-        assert np.allclose(
-            total_liquidity, target_liquidity, atol=target_liquidity * 0.05
-        ), f"test_trade.run_base_lp_test: ERROR: {target_liquidity=} does not equal {total_liquidity=}"
+        if config.simulator.init_lp:
```
Comment (dpaiton) [inline tests/test_trade.py:96]: this was my bad -- good catch

File: tests/utils/test_time_utils.py [hunk 27]
```diff
@@ -23,10 +23,13 @@ def test_current_datetime(self):
...
+        before = now - time_delta
+        after = now + time_delta
 
         assert (
-            now < test_time < (now + datetime.timedelta(milliseconds=100))
-        ), f"Unexpected time value {test_time} should be close to {now}"
+            before < test_time < after
```
Comment (dpaiton) [inline tests/utils/test_time_utils.py:31]: This is testing something different now, right? Why did you make the change?

File: tests/test_logging.py (hunk 12 above)
Comment (wakamex) [reply tests/test_logging.py:63]: it gives an error on windows

**Review by wakamex — COMMENTED** [review]
this reverts another previous incorrect attempt at a fix

File: tests/test_logging.py (hunk 12 above)
Comment (wakamex) [reply tests/test_logging.py:63]: the error is
```
>       os.remove(file_loc)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\code\\elfpy\\.logging\\test_sim.log'
```

File: src/elfpy/simulators.py (hunk 18 above)
Comment (wakamex) [reply src/elfpy/simulators.py:284]: added in validate_custom_parameters() but needs testing, left in as a TODO

File: src/elfpy/simulators.py (hunk 19 above)
Comment (wakamex) [reply src/elfpy/simulators.py:399]: I guess depends whether we want to treat init_lp as an external force or as just another agent. if closer to the latter, i think we want him to act last, since he "opens the door" on initialising liquidity, *then some things happen*, it makes sense to me we'd also ask him to "shut the door" on the way out. then may also want to check if he lost money, or otherwise treat him like any other agent. otherwise, if it's just an external force we don't care about, he doesn't even need to act at the end. but i think i prefer treating him equally.

File: src/elfpy/strategies/single_lp.py (hunk 20 above)
Comment (wakamex) [reply src/elfpy/strategies/single_lp.py:15]: ah that makes the difference clear, thanks. i was mostly trying to remove __init__ for brevity, which again results in screwing things up 🤦‍♂️

File: tests/utils/test_time_utils.py (hunk 27 above)
Comment (wakamex) [reply tests/utils/test_time_utils.py]: now < time_time seemed like a spurious test to me anyway, since both are initialized the exact same way

File: src/elfpy/simulators.py (hunk 18 above)
Comment (wakamex) [reply src/elfpy/simulators.py:304]: added testing, seems to work

File: src/elfpy/agent.py (hunk 3 above)
Comment (wakamex) [reply src/elfpy/agent.py]: removed, not needed any more

File: src/elfpy/strategies/single_lp.py (hunk 11 above)
Comment (wakamex) [reply src/elfpy/strategies/single_lp.py]: added docstrings everywhere

**Comment by wakamex** [comment]
not needed, incorporating ideas into other PRs
